### ISSUE TITLE
CLI does not work with import syntax in functions using esbuild

### ISSUE THREAD

**Issue by timroes** [netlify/cli#2068]
**Describe the bug**

I've switched to the new [`esbuild` bundling](https://www.netlify.com/blog/2021/04/02/modern-faster-netlify-functions/) and was trying to rewrite functions to use ES6 module syntax instead of CommonJS. Everything starts fine, but as soon as I am trying to execute a HTTP call against the function endpoint it will fail and the logs show the following error:

```
Request from ::1 GET /.netlify/functions/some-api
{"level":"error","message":"End - Error:"}
{"errorMessage":"Cannot use import statement outside a module","errorType":"SyntaxError","stackTrace":["ort axios from 'axios';","^^^","","taxError: Cannot use import statement outside a module","wrapSafe (internal/modules/cjs/loader.js:979:16)","Module._compile (internal/modules/cjs/loader.js:1027:27)","Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)","Module.load (internal/modules/cjs/loader.js:928:32)","Function.Module._load (internal/modules/cjs/loader.js:769:14)","Module.require (internal/modules/cjs/loader.js:952:19)","require (internal/modules/cjs/helpers.js:88:18)","Object._executeSync (/home/timroes/projects/persol.io/node_modules/lambda-local/build/lambdalocal.js:277:26)","Object.execute (/home/timroes/projects/persol.io/node_modules/lambda-local/build/lambdalocal.js:83:22)","executeSynchronousFunction (/home/timroes/projects/persol.io/node_modules/netlify-cli/src/utils/serve-functions.js:102:15)"],"level":"error"}
Response with status 500 in 23 ms.
```

I use `netlify dev --inspect` to start the dev server. Create React App is used for the frontend code. I don't use `netlify-lambda` as an additional build step for the functions. The file loaded is under `src/server/some-api/index.js`.

**Configuration**

```
[build]
  command = "yarn test && yarn build"
  publish = "build/"

[functions]
  directory = "src/server"
  node_bundler = "esbuild"

[dev]
  command = "yarn start:cra"

# Make all netlify functions available under /api/*
[[redirects]]
  from ="/api/*"
  to = "/.netlify/functions/:splat"
  status = 200
  force = true

# Stripped out some more (unrelated) redirects and headers
```

```
  System:
    OS: Linux 5.10 Arch Linux
    Memory: 19.08 GB / 31.01 GB
    Container: Yes
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.15.5 - ~/.nvm/versions/node/v14.15.5/bin/node
    Yarn: 1.22.10 - ~/projects/persol.io/node_modules/.bin/yarn
    npm: 6.14.11 - ~/.nvm/versions/node/v14.15.5/bin/npm
  npmPackages:
    netlify-cli: ^3.14.0 => 3.14.0 
```

**Expected behavior**

Executing `netlify dev --inspect` should allow running functions using ES6 modules (at least that was my assumption from the above mentioned blog post).

**Comment by eduardoboucas** [comment]
Hi @timroes! Thanks for reporting this. We still have some work to do on the new bundler during the beta period, which includes fully integrating it with the CLI so that `netlify dev` serves the bundled function and not the raw source.

The only temporary workaround I can think of is for you to change the extension of your function files to `.mjs`, so that modules work both on your local Node environment and on our build system. We're actively working on improving this and I'll make sure to update this issue as soon as we have news.

Thanks for trying the new bundler! I hope you're enjoying the experience, despite this initial hurdle.

**Comment by timroes** [comment]
Hi Eduardo,

thanks for the quick reply. Good to know that there is a workaround. Is this planned to be addressed until the esbuild bundling becomes the default on May 17th?

I also wonder, if it might be worth adding a quick note to the blog post you wrote mentioning, that limitation with `netlify dev` for now, since I could imagine a lot of users trying out the new build system also using the netlify CLI to run their projects, and might stumble across that?

Other than that looking forward to see esbuild bundling with some more functionality coming into effect, thanks for working on this. Though most likely, since I would like to use TypeScript also for functions (the same as the rest of the project) I'll need to setup netlify-lambda for building anyway, or is TS lambdas something that's planned to be addressed within the esbuild system somehow?

Cheers,
Tim

**Comment by eduardoboucas** [comment]
> Is this planned to be addressed until the esbuild bundling becomes the default on May 17th?

I don't have a definitive answer to this right now, but my expectation is that it will be available before May 17th, yes. I'll share more details when I have them.

> is TS lambdas something that's planned to be addressed within the esbuild system somehow

That is already possible! If you place a `.ts` file in your functions directory, it should work automatically.

You'll still have the `netlify dev` issue unfortunately, but this is definitely the direction we're headed.

**Comment by erezrokah** [comment]
To add more context to the issue itself, the CLI uses [`lambda-local`](https://github.com/ashiina/lambda-local) to [execute](https://github.com/netlify/cli/blob/1a963bfa52ab4c02389684008ce971f298e5d27e/src/utils/serve-functions.js#L102) functions which [dynamically requires](https://github.com/ashiina/lambda-local/blob/e1fccdac895d4ca45804d48f1d1d79116eac1f06/src/lambdalocal.ts#L238) the function.

We could:
1. Try and find a way to make this work with `lambda-local`
2. Serve the bundled version of the function (possible with source map support for better debugging). 

> At some point we would need to find a way to serve es6 modules locally, since if the function runtime is Node.js we might keep the es6 modules syntax and not transpile it to commonJS.

**Comment by eduardoboucas** [comment]
Hi all. Both ES Modules and TypeScript should now work with `netlify dev` as of version 3.17.0 of the CLI.

If you still encounter any issues at all, please add a comment and I'll reopen the issue.

Thanks!

**Comment by MentalGear** [comment]
Thx for the update. Was excited to try this out, but found out that only npm packages are supported.
Would be great if it would just work out-of-the box for normal file imports ('/utils/cleanArray.js')  as well, e.g. it would scan for local imports and add/bundle them automatically into a single file. (instead of having to use long workarounds like [this](https://answers.netlify.com/t/using-require-to-include-a-relative-module-in-netlify-functions-on-node/4177/17))

That would be great DX - I think the main selling point of Netlify, right ? :)

**Comment by erezrokah** [comment]
Hi @MentalGear, do you mind sharing a reproduction for the issue you're experiencing - possibly in a public repo? 
This should work with es6 syntax regardless if it's coming from an npm package or imported locally (we even have an es6 syntax [test case](https://github.com/netlify/cli/blob/a2799075df2b61e1653ead3f764ea09b8c5b8e4d/tests/command.dev.test.js#L1510)).

**Comment by MentalGear** [comment]
Hi Erez,
I'll try to get you a reduced case, but for now, here's the stacktrace:
```
{"level":"error","message":"End - Error:"}
{"errorMessage":"Cannot use import statement outside a module","errorType":"SyntaxError","stackTrace":["ort { cleanArray } from 'utils/cleanArray.js'","^^^","","taxError: Cannot use import statement outside a module",
"Object.compileFunction (vm.js:344:18)","wrapSafe (internal/modules/cjs/loader.js:1106:15)",
"Module._compile (internal/modules/cjs/loader.js:1140:27)","Object.Module._extensions..js (internal/modules/cjs/loader.js:1196:10)",
"Module.load (internal/modules/cjs/loader.js:1040:32)",
"Function.Module._load (internal/modules/cjs/loader.js:929:14)",
"Module.require (internal/modules/cjs/loader.js:1080:19)","require (internal/modules/cjs/helpers.js:72:18)",
"Object._executeSync (/Users/Tom/.config/yarn/global/node_modules/lambda-local/build/lambdalocal.js:277:26)",
"Object.execute (/Users/Tom/.config/yarn/global/node_modules/lambda-local/build/lambdalocal.js:83:22)"],
"level":"error"}
```

**Comment by timroes** [comment]
@MentalGear Could it be that inside that functions file where you're trying to import (and this errors happens) you accidentally forgot to have an export of the `handler` function? The error sounds like it's happening outside a detected module, which would happen if your file does not export anything.

**Comment by MentalGear** [comment]
Hey guys, thanks for your support!
Turns out, it was some weird npm/node bug. I ra  'npm update -g netlify-cli', I tried the new import statement. However, it turned out that netlify-cli never got updated and was still at 3.14. Then I noticed that the npm command was also suddenly gone, along with the binary ... Took me some hours and switched from the brew node version to n.

Happy to report that modern es import's working now as it should. 
`import { cleanArray } from '@/utils/cleanArray.js' // needed to add @ to make it work on Nuxt `
Great DX - thx for your work on this!
